### ISSUE TITLE
v0.9.5 GRID ntuple productions

### ISSUE THREAD

**Issue by yipengsun** [umd-lhcb/lhcb-ntuples-gen#87]
Here we list all GRID ntuple productions for `v0.9.5`.

## `cutflow_data`, data, 2016 MD

- [x] Production date: Sep 23, 2021 - Oct 08, 2021
- Submission script: [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/850cd3a6b8041ba66215a5ad69ef04e3789ef139/run2-rdx/jobs/21_09_23-cutflow_data-2016.sh#L1)
- Verification: [this post](https://github.com/umd-lhcb/lhcb-ntuples-gen/issues/87#issuecomment-938619155)

This was originally meant to be a `std` production but was produced as `cutflow_data` due to Yipeng's negligence. The production is sub-optimal but should be sufficient for rebuilding our run 2 data template.

The differences between `cutflow_data` and `std` are:

- `cutflow_data` doesn't have trigger filtering, so there will be more events. But we apply the trigger cuts offline anyway, so this won't be a problem
- Missing wrong-sign trees. 

### Notes

- First time run on `Stripping v28r2` (previously was on `Stripping v28r1`) and now we have ~4k subjobs, instead of just ~2k. This suggests that latest stripping divides the data more finely.
- Had a problem with some dataset, which was only available in `Beijing` and `CNAF.it`. `Beijing` kept failing so ended up getting blocked. Re-run the job at `CNAF.it` twice. The first time no ntuple was produced. The second time it produced fine.
- Job output checker appears to be working
- For Stripping v28r2, it is known that there's a small increase in number of reconstructed events. See [this email thread](https://groups.cern.ch/group/lhcb-davinci/Lists/Archive/Flat.aspx?RootFolder=%2fgroup%2flhcb%2ddavinci%2fLists%2fArchive%2fS28%20and%20S28r2%20compatibility&FolderCTID=0x01200200B795ECDBD8943F4E90E5F02A017196C9)

## `std`, data, 2011 MD

- [x] Production date: Oct 07, 2021 - Oct 13, 2021
- Submission script: [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/2ee7ccc84798a929520de35b25f7c1ba92d80158/run1-rdx/jobs/21_10_07-std-2011.sh#L1)

## `mc`, Sig and Norm, 2016 MD-MU, FullSim

The MC IDs are listed at [here](https://umd-lhcb.github.io/lhcb-ntuples-gen/data/data_sources/#signal_1) and [here](https://umd-lhcb.github.io/lhcb-ntuples-gen/data/data_sources/#normalization_1)

- [x] Production date: Oct 08, 2021 - Oct 14, 2021
- Submission script: [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/6d49a154dd3044db3bddd598dbb4b37c2b3a026d/run2-rdx/jobs/21_10_08-fullsim_sig_nor_reprod.sh#L1)

## `mc`, `DDX` and `D**s`, 2016 MD-MU, FullSim

MC IDs [here](https://umd-lhcb.github.io/lhcb-ntuples-gen/data/data_sources/#d_s_1) and [here](https://umd-lhcb.github.io/lhcb-ntuples-gen/data/data_sources/#ddx_1)

- [x] Production date: Oct 19, 2021 - Oct 29, 2021
- Submission script: [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/87e2f894074d1d12acf373c78f9fc0434be5d64f/run2-rdx/jobs/21_10_08-fullsim_DD-Dstst_s.sh)

## `mc`, `D*+TauNu`, 2012 MD-MU, FullSim, Sim08a

MC IDs: `11574010`

- [x] Production date: Oct 15, 2021 - Oct 25, 2021
- Submission script: [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/d02e2978b183e66ae0fd3a7398a3f74349837a1f/run1-rdx/jobs/21_10_15-fullsim_DstTauNu.sh)
- Ganga job id: `146, 148`
- Verification: [this link](https://github.com/umd-lhcb/lhcb-ntuples-gen/issues/87#issuecomment-950411484)

## `mc`, `D**`, 2016 MD-MU, FullSim

MC IDs [here](https://umd-lhcb.github.io/lhcb-ntuples-gen/data/data_sources/#d_1) and [here](https://umd-lhcb.github.io/lhcb-ntuples-gen/data/data_sources/#d-rightarrow-d-pi-pi_1).

- [x] Production date: Oct 15, 2021 - now
- Submission script: [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/ad48a36d6f9396595a6e2ae5c64a8acdb2ebe173/run2-rdx/jobs/21_10_15-fullsim_Dstst_reprod.sh)
- Verification: [this post](https://github.com/umd-lhcb/lhcb-ntuples-gen/issues/87#issuecomment-944610334)
- Ganga job id: `149-169`

## `mc`, `D* -> Tau/Mu Nu`, 2016 MD-MU, tracker-only

MC IDs: `11574011, 11574021`

- [ ] Production date: Oct 16, 2021 - now
- Submission script: [here](https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/7ddf4dba196a7c38819552ccf0137ef4fb73f97a/run2-rdx/jobs/21_10_16-tracker_only_DstTau_DstMu.sh)
- Ganga job id: `170-173`

## Postprocess of GRID ntuples

I typically remove some branches during this process. Use the `scripts/haddcut.py`:

```
haddcut.py <output_ntp> <input_ntp> -c <YAML_config>
```

For the `<YAML_config>`:

- data: https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/master/postprocess/skims/rdx_data.yml
- FullSim: https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/master/postprocess/skims/rdx_mc_fullsim.yml
- Tracker-only: https://github.com/umd-lhcb/lhcb-ntuples-gen/blob/master/postprocess/skims/rdx_mc.yml

**Comment by yipengsun** [comment]
Also, can you add your actual job submission script to `run2-rdx/jobs`? You can start with one of the existing script and just tweak the parameters.

**Comment by yipengsun** [comment]
Also, after `hadd` the ntuples on lxplus, I typically drop some branches offline to make the ntuples smaller (because they can be HUGE). I'll document that procedure later, but just keep this in mind that you need to run another script before doing the `git annex add` procedure.

**Comment by afernez** [comment]
I tried running a job overnight last night to do one of the `DDX` modes, just to make sure I had things configured correctly. Weirdly, ganga stopped before running the jobs with an error message 
`ERROR    SandboxError: File '/afs/cern.ch/user/a/afernez/lhcb-ntuples-gen/run2-rdx/weights_soft.xml' does not exist.`
This file does exist... so I'm not sure I understand what went wrong. I'll look into it now, and then I'll try to run the same job again, and then I'll do all of the `DDX` and `D**s`.

Also, yeah, I won't `git annex add` the ntuples right away after merging; I'll wait to run your script that cuts out unneeded branches.

**Comment by afernez** [comment]
Ok, submitted the jobs for all 2016 fullsim `DDX` and `D**s` md and mu. I'll commit the job script and update the top post.

**Comment by yipengsun** [comment]
### Verification for `cutflow_data`, data, 2016 MD
```
Verifying output for Job 130...
subjob 0: ntuple has a size of 5 KiB, which is too small!
subjob 15: ntuple has a size of 5 KiB, which is too small!
subjob 16: ntuple has a size of 5 KiB, which is too small!
subjob 2: ntuple has a size of 5 KiB, which is too small!
subjob 20: ntuple has a size of 5 KiB, which is too small!
subjob 21: ntuple has a size of 5 KiB, which is too small!
subjob 23: ntuple has a size of 5 KiB, which is too small!
subjob 26: ntuple has a size of 5 KiB, which is too small!
subjob 27: ntuple has a size of 5 KiB, which is too small!
subjob 28: ntuple has a size of 5 KiB, which is too small!
subjob 3: ntuple has a size of 5 KiB, which is too small!
subjob 30: ntuple has a size of 5 KiB, which is too small!
subjob 38: ntuple has a size of 5 KiB, which is too small!
subjob 39: ntuple has a size of 5 KiB, which is too small!
subjob 4: ntuple has a size of 5 KiB, which is too small!
subjob 41: ntuple has a size of 5 KiB, which is too small!
subjob 42: ntuple has a size of 5 KiB, which is too small!
subjob 45: ntuple has a size of 5 KiB, which is too small!
subjob 46: ntuple has a size of 5 KiB, which is too small!
subjob 47: ntuple has a size of 5 KiB, which is too small!
subjob 51: ntuple has a size of 5 KiB, which is too small!
subjob 58: ntuple has a size of 5 KiB, which is too small!
subjob 62: ntuple has a size of 5 KiB, which is too small!
subjob 63: ntuple has a size of 5 KiB, which is too small!
subjob 69: ntuple has a size of 5 KiB, which is too small!
subjob 7: ntuple has a size of 5 KiB, which is too small!
Job 130 output verification failed with 26 error(s).
```

Out of 4264 subjobs, 26 produced ntuples don't contain any candidate.

In the end I set the minimal file size to 1 KiB.

The merge failed with the following errors:
```
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
2_90000000_SEMILEPTONIC.DST.root).
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
2_90000000_SEMILEPTONIC.DST.root).
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
2_90000000_SEMILEPTONIC.DST.root).
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
2_90000000_SEMILEPTONIC.DST.root).
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
2_90000000_SEMILEPTONIC.DST.root).
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
2_90000000_SEMILEPTONIC.DST.root).
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
2_90000000_SEMILEPTONIC.DST.root).
Warning in <TTree::CopyEntries>: The output TTree (DecayTree) must be associated with a writable directory (TupleB0 in eos/stash/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_Reco16_Stripping28r
```

I downloaded the full job output, and `hadd`ed them locally successfully:
```
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/GetIntegratedLuminosity
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleBminus
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleB0
hadd Opening the next 923 files
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/GetIntegratedLuminosity
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleBminus
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleB0
hadd Opening the next 923 files
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/GetIntegratedLuminosity
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleBminus
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleB0
hadd Opening the next 923 files
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/GetIntegratedLuminosity
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleBminus
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleB0
hadd Opening the next 574 files
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/GetIntegratedLuminosity
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleBminus
hadd Target path: downloads/Dst_D0--21_09_23--cutflow_data--LHCb_Collision16_Beam6500GeV-VeloClosed-MagDown_Real_Data_eco16_Stripping28r2_90000000_SEMILEPTONIC.DST.root:/TupleB0
```

**Comment by yipengsun** [comment]
@afernez I decided to put verification into separate comments, and just provide link in the top post, to avoid making top post super long.

**Comment by yipengsun** [comment]
Production for large tracker-only 2016 MC for both polarities for `D* -> Tau/Mu Nu` has been submitted.

**Comment by afernez** [comment]
Ok, I think ganga is cooperating enough for me now to resubmit the `DD`+`D**s` 2016 MU-MD FullSim MC jobs. I've updated the top post with the new date. I'm hopeful these jobs will all run correctly now... It is unclear to me what the issue has been.

**Comment by yipengsun** [comment]
There's a main difficulty in merging the large TO ntuple: the EOS dismounts within 1 day:

I have tried to either mount the EOS via `sshfs` or merging directly on lxplus, both times the merger errored out with a _file not exist_ error, I checked and confirmed that the EOS storage has been dismounted. I think it is only mounted for about a day before it dismounts.

I have resorted back to copy the full files to my local machine and merge locally. Unfortunately this puts a huge stress on my disk space and I'm trying to deleted unneeded files to have enough space for the ntuples.
